ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java

```diff
@@ -171,7 +201,19 @@ public long writeTo(int index, WritableByteChannel channel, long relativeOffset,
     long startOffset = options.getOffset() + relativeOffset;
     long sizeToRead = Math.min(maxSize, options.getMessageInfo().getSize() - relativeOffset);
     logger.trace("Blob Message Read Set position {} count {}", startOffset, sizeToRead);
-    long written = options.getChannel().transferTo(startOffset, sizeToRead, channel);
+    long written = 0;
+    if (options.getPreFetchedDataRelativeOffset() != -1) {
+      ByteBuffer buf = options.getPreFetchedData();
+      buf.flip();
+      long bufStartOffset = relativeOffset - options.getPreFetchedDataRelativeOffset();
+      buf.limit((int)(bufStartOffset + sizeToRead));
+      buf.position((int)(bufStartOffset));
+      while (buf.hasRemaining()) {
```

Comment:
I believe the caller handles retries until fully done - we should check that

Reply:
fixed.